src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1309 @@
+    //
+    // Parameters:
+    //   comp   - Compiler instance
+    //   lclNum - Local num for this struct local
+    //   replacements - [out] Pointer to vector to create and insert replacements into
+    //
+    void PickPromotions(Compiler* comp, unsigned lclNum, jitstd::vector<Replacement>** replacements)
+    {
+        if (m_accesses.size() <= 0)
+        {
+            return;
+        }
+
+        assert(*replacements == nullptr);
+        for (size_t i = 0; i < m_accesses.size(); i++)
+        {
+            const Access& access = m_accesses[i];
+
+            if (access.AccessType == TYP_STRUCT)
+            {
+                continue;
+            }
+
+            if (!EvaluateReplacement(comp, lclNum, access))
+            {
+                continue;
+            }
+
+#ifdef DEBUG
+            char buf[32];
+            sprintf_s(buf, sizeof(buf), "V%02u.[%03u..%03u)", lclNum, access.Offset,
+                      access.Offset + genTypeSize(access.AccessType));
+            size_t len  = strlen(buf) + 1;
+            char*  bufp = new (comp, CMK_DebugOnly) char[len];
+            strcpy_s(bufp, len, buf);
+#endif
+            unsigned   newLcl = comp->lvaGrabTemp(false DEBUGARG(bufp));
+            LclVarDsc* dsc    = comp->lvaGetDesc(newLcl);
+            dsc->lvType       = access.AccessType;
+
+            if (*replacements == nullptr)
+            {
+                *replacements =
+                    new (comp, CMK_Promotion) jitstd::vector<Replacement>(comp->getAllocator(CMK_Promotion));
+            }
+
+            (*replacements)->push_back(Replacement(access.Offset, access.AccessType, newLcl));
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // EvaluateReplacement:
+    //   Evaluate legality and profitability of a single replacement candidate.
+    //
+    // Parameters:
+    //   comp   - Compiler instance
+    //   lclNum - Local num for this struct local
+    //   access - Access information for the candidate.
+    //
+    // Returns:
+    //   True if we should promote this access and create a replacement; otherwise false.
+    //
+    bool EvaluateReplacement(Compiler* comp, unsigned lclNum, const Access& access)
+    {
+        weight_t countOverlappedCallsWtd                 = 0;
+        weight_t countOverlappedReturnsWtd               = 0;
+        weight_t countOverlappedRetbufsWtd               = 0;
+        weight_t countOverlappedAssignmentDestinationWtd = 0;
+        weight_t countOverlappedAssignmentSourceWtd      = 0;
+
+        bool overlap = false;
+        for (const Access& otherAccess : m_accesses)
+        {
+            if (&otherAccess == &access)
+                continue;
+
+            if (!otherAccess.Overlaps(access.Offset, genTypeSize(access.AccessType)))
+            {
+                continue;
+            }
+
+            if (otherAccess.AccessType != TYP_STRUCT)
+            {
+                return false;
```

Comment:
Disqualifying these replacements is the simple thing to do for now. In the future we can allow overlapping accesses using a few possible strategies (e.g. inserting `GT_BITCAST` when the lengths are the same, or using bitwise operators when they aren't).